lib/plugins/rest.js

```diff
@@ -189,23 +189,62 @@ exports.register = function (server, options, next) {
 						uri      : request.payload.uri,
 						revision : request.payload.revision
 					});
-
-					return instances.updateInstance(instance)
-					.then(function (response) {
-						reply(response);
-					})
-					.catch (function (error) {
-						server.log([ "error", "rest" ], error.message);
-						if (error.toString().match(/\bnot exist\b/)) {
-							reply(Boom.notFound("This instance does not exist."));
-						}
-						else if (error.toString().match(/\bconcurrency\b/)) {
-							reply(Boom.conflict("Conflict error. Ensure you have the latest document version"));
+					switch (instance.state) {
+						case ("terminating") : {
+							return instances.updateInstance(instance)
+							.then(function (response) {
+								reply(response);
+								return awsAdapter.terminateInstances(response.id);
+							})
+							.then(function () {
+								return instances.getInstance({ id : request.params.id });
+							})
+							.then(function (response) {
+								instance = new Instance({
+									id       : response.id,
+									ami      : response.ami,
+									type     : response.type,
+									state    : "terminated",
+									uri      : null,
+									revision : response.revision
+								});
+								return instances.updateInstance(instance);
+							})
+							.catch (function (error) {
```

Comment:
The problem is that the error handling is happening in the same promise chain that the `reply()` call happens in. It may be an obscure edge case, but consider what would happen if the `reply()` call on line 196 succeeds but then the database goes down. In this case the code will attempt to call `reply()` again on line 222. I don't think that hapi will allow this and it may very well obscure what the actual problem is.